### DISCUSSION TITLE
Formatter: `np.array` formatting

### DISCUSSION THREAD

**Discussion by MichaReiser** [astral-sh/ruff#8452] [General]
Broader adoption of ruff's formatter in the scientific community is the `np.array` (see [this thread in the Beta Feedback discussion](https://github.com/astral-sh/ruff/discussions/7310#discussioncomment-7266904)). The solution used by the community today is to suppress the formatting with `fmt:off`/`fmt:on` or `fmt:skip`. However, Ruff's lack of support for expression-level suppressions results in unnecessarily large suppressed ranges (see [this comment in the Suppression Comments discussion](https://github.com/astral-sh/ruff/discussions/6338#discussioncomment-7434505)).

[The thread](https://github.com/astral-sh/ruff/discussions/7310#discussioncomment-7266904) discusses two options:

* Disabling formatting for arrays passed to `np.array` 
* Improving formatting for arrays passed to `np.array`, but without an explicit proposal. 

I prefer not to disable formatting because it means the formatter fails its main goal --- guaranteed consistent code formatting. It also takes the pressure from us to come up with a style that works well for the majority of cases, leaving the ecosystem in a worse state overall. 

I want to use this issue to discuss a concrete style for improving `np.array` formatting. I have never used `np.array`, so I'll make incorrect assumptions. This is why I would love to hear feedback on the proposal from people who have used `np.array`. Please also correct me if the goal needs refinement.

**Goal**: Improve formatting of the object passed to `np.array` calls if it is a **two** dimensional array. 
**Non-Goal**: Improve formatting for one or three and more dimensional arrays. One-dimensional list formatting may need an overhaul overall, but this is not specific to `np.array`

### New Style Applicability

When should the new style apply (rather than using the generic list formatting)

* A `CallExpression` where the function is an `AttributeExpr(name=np, attribute=array)` 
* The first argument is a two-dimensional list
	* All rows have the same number of cells 
	* The rows are not empty
	* if `skip-magic-trailing-comma=false`: No row has a trailing comma (indicating that the values should be split on their own 
line)
	* if `skip-magic-trailing-comma=true`: Ignore the style regardless whether a trailing comma is present or not
* All values are constant numbers
* The cells (values) have no comments  

It is okay for:

* The two-dimensional array or the row to have leading or trailing comments

**Open Questions**

* Are there other common import names for numpy other than `np`?
* Do we need to support other methods?
* Should we apply a length limit to numbers (max 10 characters?) (No?)
* Should we apply this style to all two-dimensional, number-only arrays? (No?)

## Proposed style

* Each row gets formatted on its own line
* Cells are kept on a single line, even at the cost of exceeding the configured line length. 
* Cells are padded to the width of the largest value 

Example:

```python
    img_ = np.array([
		 [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0],
         [  0,   0,   0,   0,   0,  82, 180, 236, 255, 236, 180,  82,   0,   0,   0,   0,   0],
         [  0,   0,   0,   0, 189, 172,  74,  18,   0,  18,  74, 172, 189,   0,   0,   0,   0],
         [  0,   0,   0, 229,  25,   0,   0,   0,   0,   0,   0,   0,  25, 229,   0,   0,   0],
         [  0,   0, 189,  25,   0,   0,   0,   0,   0,   0,   0,   0,   0,  25, 189,   0,   0],
         [  0,  82, 172,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 172,  82,   0],
         [  0, 180,  74,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  74, 180,   0],
         [  0, 236,  18,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  18, 236,   0],
         [  0, 255,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 255,   0],
         [  0, 236,  18,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  18, 236,   0],
         [  0, 180,  74,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  74, 180,   0],
         [  0,  82, 172,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 172,  82,   0],
         [  0,   0, 189,  25,   0,   0,   0,   0,   0,   0,   0,   0,   0,  25, 189,   0,   0],
         [  0,   0,   0, 229,  25,   0,   0,   0,   0,   0,   0,   0,  25, 229,   0,   0,   0],
         [  0,   0,   0,   0, 189, 172,  74,  18,   0,  18,  74, 172, 189,   0,   0,   0,   0],
         [  0,   0,   0,   0,   0,  82, 180, 236, 255, 236, 180,  82,   0,   0,   0,   0,   0],
         [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0]
	])
```

Note: The example assumes the preview "hug list preview style".

### Alternatives

Instead of padding all cells to the same width, compute the padding for each column. 

```python
    img_ = np.array([
        [0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0],
        [0,   0,   0,   0,   0,  82, 180, 236, 255, 236, 180,  82,   0,   0,   0,   0, 0],
        [0,   0,   0,   0, 189, 172,  74,  18,   0,  18,  74, 172, 189,   0,   0,   0, 0],
        [0,   0,   0, 229,  25,   0,   0,   0,   0,   0,   0,   0,  25, 229,   0,   0, 0],
        [0,   0, 189,  25,   0,   0,   0,   0,   0,   0,   0,   0,   0,  25, 189,   0, 0],
        [0,  82, 172,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 172,  82, 0],
        [0, 180,  74,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  74, 180, 0],
        [0, 236,  18,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  18, 236, 0],
        [0, 255,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 255, 0],
        [0, 236,  18,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  18, 236, 0],
        [0, 180,  74,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,  74, 180, 0],
        [0,  82, 172,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 172,  82, 0],
        [0,   0, 189,  25,   0,   0,   0,   0,   0,   0,   0,   0,   0,  25, 189,   0, 0],
        [0,   0,   0, 229,  25,   0,   0,   0,   0,   0,   0,   0,  25, 229,   0,   0, 0],
        [0,   0,   0,   0, 189, 172,  74,  18,   0,  18,  74, 172, 189,   0,   0,   0, 0],
        [0,   0,   0,   0,   0,  82, 180, 236, 255, 236, 180,  82,   0,   0,   0,   0, 0],
        [0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0]
])
```

### Considerations

* Applying padding to the cells (the same for all columns or column-wise) has the disadvantage that changing the width of a single number may result in a large diff when it results in a changed padding (for all cells)
* The proposed style intentionally trades "better" formatting against respecting the configured line length.

**Comment by peterjc** [comment]
Personally I would favour computing the minimum width for each column, as our arrays for Biopython are mainly in test cases, which rarely change.

**Comment by peterjc** [comment]
Also I like the suggested layout:

```python
np.array([
   # rows here
])
```

i.e. Putting the ( and [ on the same line not to waste vertical space.

**Reply by jack-mcivor** [reply]
What if there are other arguments to the constructor? Like
```python
np.array([
    [0, 0, 0, 0],
    [0, 0, 0, 0]
], dtype=np.uint8)
```

**Reply by MichaReiser** [reply]
I suggest to format it the same as other function calls where one argument "expands" over multiple lines:

```python
np.array(
	[
	    [0, 0, 0, 0],
	    [0, 0, 0, 0],
	], 
	dtype=np.uint8,
)
```

**Reply by peterjc** [reply]
That'd be OK in my personal opinion, however I would prefer less vertical space and less indentation (more like the numpy repr). In my own usage an explicit ``dtype=`` is quite common.

**Comment by peterjc** [comment]
Answering the Open Questions:

* Q: Are there other common import names for numpy other than np?
  - Yes, ``numpy`` (since not everyone uses the as ``np`` convention)
* Q: Do we need to support other methods?
  - Possibly ``np.matrix``?
* Q: hould we apply a length limit to numbers (max 10 characters?) (No?)
  - No, I would focus on any line length limit instead
* Q: Should we apply this style to all two-dimensional, number-only arrays? (No?)
  - Unsure, do you mean things like lists of lists even when not passed to ``np.array``? Note these might have mixed column data types...

**Reply by MichaReiser** [reply]
* > No, I would focus on any line length limit instead
  * Do you mean the configured line-length limit. The proposed style would intentionally ignore the line length limit to ensure that it can fit the rows, regardless of their width. 
* > Unsure, do you mean things like lists of lists even when not passed to np.array? Note these might have mixed column data types...
   * I was thinking of all lists of lists that only contain number values.

**Reply by peterjc** [reply]
Oh, OK. I can see some reasons for decoupling the array layout from the global line length limit. But I would set the limit in terms of number of characters required (or line width), not number of elements (column count). This is partly because integers vs floats can make a big difference, and also arrays could be defined at several levels of indentation.

I think lists of lists vs lists of lists of numbers is a marginal call (not something I think affects my code or projects much). I have no strong view here.

**Reply by MichaReiser** [reply]
> Oh, OK. I can see some reasons for decoupling the array layout from the global line length limit. But I would set the limit in terms of number of characters required (or line width), not number of elements (column count)

The idea isn't to introduce a new option but instead ignore the line-length entirely. It's the only way how we can guarantee that each row is on its own line.

**Reply by peterjc** [reply]
OK - no limits so each row is one line does make sense as a default. Very large arrays/rows are going to be problematic anyway so the author may opt out of automatic formatting here.

**Reply by MichaReiser** [reply]
Yeah. I don't have experience on how common the very large rows is or if it would still be preferred to have them as single row than the one line per row formatting (which gets very long as well). That's where `fill` layout would be interesting that simply tries to put as many cells as possible onto a single line. But I think that's not what `np.array` users generally want? (at least from the suppression comments that I've seen). 

Users can enforce the one-line per cell layout by adding a trailing semicolon to a row:

```python
np.array([
	[1, 2,]
])

# becomes

np.array([
	[
		1, 
		2,
	]
])
```

**Comment by jack-mcivor** [comment]
Should this also handle arrays >2D? I guess the innermost list can be formatted as rows?

Note that the numpy array representation does something similar by default (each row on a new line, columns right justified & all padded to the width of the maximum value)
```
>>> np.array([[[0,0,0],[0,0,0]],[[0,0,0],[0,10,100]]])
array([[[  0,   0,   0],
        [  0,   0,   0]],

       [[  0,   0,   0],
        [  0,  10, 100]]])
```
The relevant function I think is https://numpy.org/doc/stable/reference/generated/numpy.array_repr.html, which has a lot of options and handling for different data types (see https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html#numpy.set_printoptions)

**Reply by peterjc** [reply]
Maybe the style should be based based on the numpy repr (which is more compact than the examples given earlier)?

**Reply by MichaReiser** [reply]
Interesting. The challenge I see with formatting is that we can't pass options. So it would be necessary to pick a good default. 
Is the 2-dimension style different from what I proposed? How common are these 2+ dimension arrays?

**Reply by jack-mcivor** [reply]
> Is the 2-dimension style different from what I proposed?

The next line matches the indentation of the previous. Compare 
```
>>> np.array([[[0,0,0],[0,0,0]],[[0,0,0],[0,10,100]]])
array([[[  0,   0,   0],
        [  0,   0,   0]],

       [[  0,   0,   0],
        [  0,  10, 100]]])

# add a bunch of extra outer dimensions
>>> np.array([[[[[[0,0,0],[0,0,0]],[[0,0,0],[0,10,100]]]]]])
array([[[[[[  0,   0,   0],
           [  0,   0,   0]],

          [[  0,   0,   0],
           [  0,  10, 100]]]]]])
```
This is sort of against the black style isn't it? It seems semantically like
```python
do_thing(first_arg,
         second_arg)
```
You could ignore "coalescing" the brackets and just go for a style like
```
>>> np.array([[[0,0,0],[0,0,0]],[[0,0,0],[0,10,100]]])
array(
    [
        [
            [  0,   0,   0],
            [  0,   0,   0]
        ],
        [
            [  0,   0,   0],
            [  0,  10, 100]
        ]
    ]
)
```

> How common are these 2+ dimension arrays?

I come across them quite a bit I guess? The nd structure is definitely a core concept in numpy. We could check the source of some popular projects to see how often they are written out - eg. scipy/numpy

**Comment by jl-wynen** [comment]
Numpy isn't the only array library out there, although it is the most widespread one. For example, I work on [Scipp](https://github.com/scipp/scipp) which adds a few extra properties to arrays. We face similar issues with formatting 2d arrays.

So I was wondering if it is possible to find a common solution. Two solution that come to my mind are

- A config option to choose what functions trigger array formatting. E.g., `array-formatting = ["np.array", "sc.array"]`.
- Implementing
> - Should we apply this style to all two-dimensional, number-only arrays? (No?)

The latter might provide a better out-of-the-box experience. Conceptually, I don't see why a numpy array constructed from a list would be treated differently from a plain list. (Of course there may be technical reason for why detecting lists of lists of numbers might not be feasible.)

For context, in Scipp, creating an array might look like this:
```python
sc.array(
    dims=['x', 'y'],
    values=[[1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5]],
    unit='m',
)
```
and potentially even with two lists:
```python
sc.array(
    dims=['x', 'y'],
    values=[[1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5]],
    variances=[[1, 2, 3, 4, 5],
               [1, 2, 3, 4, 5],
               [1, 2, 3, 4, 5]],
    unit='m',
)
```
In this case, all arguments are required to be keywords. And there are more functions that take lists like this as arguments.

**Reply by lagru** [reply]
> Should we apply this style to all two-dimensional, number-only arrays? (No?)

From a user perspective this approach seems to me as the most consistent one with the least amount of surprising behavior. I'd imagine an option like `--pad-nested-number-only-lists` is simpler to understand than maintaining every possible case in `array-formatting = ["np.array", "sc.array", ...]`

**Reply by MichaReiser** [reply]
> , I don't see why a numpy array constructed from a list would be treated differently from a plain list.

The main reason is that the proposed design trades keeping the code in the configured line length with keeping rows together. Keeping the code inside the configured line length (where possible) is a fundamental objective of the formatter. Removing this entirely for all multidimensional, number only-lists, may be undesired for many. Altough I don't have any data on this. 

The other challenge I see is that having multiple opening/closing parentheses on the same line is unprecedented until now and deviates from other formatting and would also be different from how multidimensional non-number lists would be formatted.

**Reply by NeilGirdhar** [reply]
@MichaReiser I think his point is that whatever you do for numpy arrays you should also do for plain lists.  Has there been any justification for deviation?

As for line length, you can still have a grid formatting that that obeys the maximum line length:
```
    img_ = np.array([
      [0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0, 0,
       0,   0],
      [0,   0,   0,   0,   0,  82, 180, 236, 255, 236, 180,  82,   0,   0, 0,
       0,   0]])
```
I don't think you should remove the line length limit even for the _grid formatting_ that you proposed.  (Or at least that's a separate proposal.)

> The other challenge I see is that having multiple opening/closing parentheses on the same line is unprecedented until now

Yes, it's a different style.  The multiple open-close parentheses are simply an example of _vertical formatting_.  Currently Ruff only supports _hanging formatting_.

And yes, I think if you're going to allow vertical formatting in some circumstances, it should be uniformly adopted rather than special-cased.  For example, vertical is superior to hanging (in my opinion) when it uses fewer lines, which happens when it induces no additional line-wraps.

**Reply by jl-wynen** [reply]
> @MichaReiser I think his point is that whatever you do for numpy arrays you should also do for plain lists. Has there been any justification for deviation?

Yes, this is what I meant.

And I agree that removing the line length limit for grid formatting may not be ideal in general. On the other hand, 2d arrays are usually easier to read with vertical scrolling than with line breaks. But I don't have enough experience to judge how large those grids typically are and whether 80 or so columns are enough.

Concerning multiple closing parentheses, this is what numpy itself does:
```python
>>> np.ones((4, 4))
array([[1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.],
       [1., 1., 1., 1.]])
```

**Comment by effigies** [comment]
While `np.array` can be special-cased, the existence of a bunch of [other array providers](https://numpy.org/neps/nep-0047-array-api-standard.html) seems to make it worth-while to think about ND-array-like literals, regardless of whether they're wrapped in `np.array()`. Here is a possible rule:

> When a nested list is detected and the constituent lists are all of the same length (it is N-dimensional, not ragged), then coalesce any opening and closing braces. When a brace closes and another opens, add a newline and align the next open brace to the level of the matching brace. For the purposes of aligning further tokens, treat the object as if it were on a single line, and do not automatically expand to multi-line mode.

```python
# 2D with trailing comma
np.array(
    [[1, 0],
     [0, 1]],
)

# 2D without trailing comma
np.array([[1, 0],
          [0, 1]])

# With dtype and trailing comma
np.array(
    [[1, 0],
     [0, 1]],
    dtype='int64',
)

# With dtype and without trailing comma
np.array([[1, 0],
          [0, 1]], dtype='int64')
```

This already gets us sensible behavior of array-like literals and keyword arguments in the same constructor. But we don't need to special case `np.array()`, positional arguments or 2D arrays. Applying this rule to keyword arguments and other constructors gives us @jl-wynen's example above:

```python
sc.array(
    dims=['x', 'y'],
    values=[[1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5],
            [1, 2, 3, 4, 5]],
    variances=[[1, 2, 3, 4, 5],
               [1, 2, 3, 4, 5],
               [1, 2, 3, 4, 5]],
    unit='m',
)
```

<details>
<summary>I know they're out-of-scope, but expand here to see the extension to 3- and 4D arrays:</summary>

```python
# 3D
array(
    [[[1., 0.],
      [0., 1.]],
     [[0., 1.],
      [1., 0.]]],
)

# np.arange(48).reshape((2, 3, 2, 4))
array(
    [[[[ 0,  1,  2,  3],
       [ 4,  5,  6,  7]],
      [[ 8,  9, 10, 11],
       [12, 13, 14, 15]],
      [[16, 17, 18, 19],
       [20, 21, 22, 23]]],
     [[[24, 25, 26, 27],
       [28, 29, 30, 31]],
      [[32, 33, 34, 35],
       [36, 37, 38, 39]],
      [[40, 41, 42, 43],
       [44, 45, 46, 47]]]],
)
```

</details>

Applying the rule to the arguments of a list literal gives us 3D array-likes almost by accident:

```python
numbers = [
    [[1, 2, 3, 4],
     [5, 6, 7, 8]],
    [[ 9, 10, 11, 12],
     [13, 14, 15, 16]],
]
```

And if we use parentheses to indicate that the top-level list is to be treated as a single ND array instead of 2 2D arrays:

```python
numbers = (
    [[[ 1,  2,  3,  4],
      [ 5,  6,  7,  8]],
     [[ 9, 10, 11, 12],
      [13, 14, 15, 16]]]
)
```

It's getting even more out of the scope of `np.array()`s, so I'll just say but not really argue for this: it would not be necessary to limit yourself to lists of lists. Lists/sets/tuples of lists/sets/tuples, where the container type and size at each level is consistent would be amenable to the same kind of alignment.

**Reply by NeilGirdhar** [reply]
@effigies I like your suggestion too.  This is just _vertical formatting_, which I proposed here: https://github.com/astral-sh/ruff/issues/8197

Honestly, I don't think this should be special-cased for arrays, but rather a better approach to formatting should be considered that takes vertical space into account.

**Reply by effigies** [reply]
I don't think it's incompatible with what you proposed, but it's not quite the same, I don't think. And I do think that it is worth special-casing for nested lists, as this is an improvement in the current style as well as under your vertical formatting proposal. If your proposal is adopted and allows this special case to be removed, that would be fine, too.

**Reply by NeilGirdhar** [reply]
> that it is worth special-casing for nested lists, as this is an improvement in the current style 

Right, but if it's an improvement for lists, it's an improvement in other places like nested generator expressions or parameter specifications.  We should be pushign for the general solution IMO.

**Comment by MichaReiser** [comment]
Thank you all for the excellent feedback. I don't know when I'll get time to look into this closer but I certainly have a better understanding for the problem than before and it might be worth trying to find a formatting that works for all multi dimensional arrays. Altough this may require more time to stabilize, which could justify special casing for scientific usages for now to remove the most common need for suppression comments in the short term.

**Comment by Sarcasm** [comment]
The `zig fmt` command has a rule:

> For array literals, regardless of trailing comma or not, if there are a subset of the total number of items in the first line of the literal, that is taken as the row size, and the array literal is formatted with all rows of that size, with a trailing comma.
> -- https://github.com/ziglang/zig/issues/1003

I wonder if the same logic can be used by ruff to format list-of-lists.
With the first row, the users can tell the formatter what to do with the remaining ones.

**Comment by randolf-scholz** [comment]
A few comments:

1. I'd also want such formatting not just to apply to `np.array` but more general to tabular-like and array-like expressions
   - More generally, nested list/dict/tuple/set expressions, when one can detect that the inner container has a constant schema across all elements of the outer container.
   - for example take a `list[list[int | str]]` if each inner list is a list expression with the same number of elements, and a the data type, like `[int, str, str]` then that woud be tabular-like. 
2. It is useful to distinguish 2 cases: 
    - **array-like data**: homogeneous n-dimensional arrays (np.ndarray, torch.Tensor, etc.) — all elements share the same data type, usually numerical (bool/int/float/complex)
    - **tabular-like data**: hereterogenous nested data (pandas.DataFrame, pyarrow.Table, etc.) with both numerical and string columns.
    
   For the former, most libraries use a a fixed padding width based on the largest element in the whole array, for the latter, column-wise padding values are more appropriate
3. If auto-detection fails, it would be great if one could force a particular formatting with a pragma like `# fmt: tabular` / `# fmt: array`.
